File: netlify/functions/lib/scrapers/the1337x.js
/**
 * 1337x scraper - HTML scrape with two-phase fetch (list -> detail page).
 */
const client = require('../client');
const { parseSeeders, parseSize, cleanName } = require('../utils');

const NAME = '1337x';
const BASE = 'https://1337x.to';

async function search(query) {
  const url = `${BASE}/search/${encodeURIComponent(query)}/1/`;
  const $ = await client.getSoup(url);
  if (!$) return [];

  const links = [];
  $('td.coll-1.name a').each((_, el) => {
    const href = $(el).attr('href') || '';
    if (href.includes('/torrent/')) links.push(BASE + href);
  });

  if (!links.length) return [];

  // Fetch detail pages concurrently (limit to 15)
  const detailPromises = links.slice(0, 15).map(async (detailUrl) => {
    try {
      const page = await client.getSoup(detailUrl);
      if (!page) return null;

      const name = cleanName(page('div.box-info-heading h1').text());
      if (!name) return null;

      const magnetTag = page('a[href^="magnet:"]').first();
      if (!magnetTag.length) return null;
      const magnet = magnetTag.attr('href');

      let sizeText = '';
      let seedersText = '0';
      page('ul.list li').each((_, li) => {
        const label = page(li).find('strong').text().toLowerCase();
        const value = page(li).find('span').text().trim();
        if (label.includes('size')) {
          sizeText = value.replace(/\(.*?\)/g, '').trim();
        } else if (label.includes('seeders')) {
          seedersText = value;
        }
      });

      return {
        name,
        size: sizeText || 'N/A',
        size_gb: parseSize(sizeText),
        seeders: parseSeeders(seedersText),
        magnet,
        provider: NAME,
      };
    } catch {
      return null;
    }
  });

  const settled = await Promise.all(detailPromises);
  return settled.filter(Boolean);
}

module.exports = { name: NAME, base_url: BASE, search };

File: scrapers/the1337x.py
"""1337x scraper - HTML scraping with two-phase fetch (list -> detail page)."""

import re
from urllib.parse import quote
from concurrent.futures import ThreadPoolExecutor, as_completed
from scrapers.base import BaseScraper
from scrapers import client, utils


class Source(BaseScraper):
    name = '1337x'
    base_url = 'https://1337x.to'

    def search(self, query):
        search_url = f'{self.base_url}/search/{quote(query)}/1/'
        soup = client.get_soup(search_url)
        if not soup:
            return []
        links = []
        for a_tag in soup.select('td.coll-1.name a[href*="/torrent/"]'):
            href = a_tag.get('href', '')
            if href:
                links.append(self.base_url + href)
        if not links:
            return []
        results = []

        def fetch_detail(url):
            page = client.get_soup(url)
            if not page:
                return None
            name_tag = page.select_one('div.box-info-heading h1')
            name = utils.clean_name(name_tag.get_text()) if name_tag else ''
            if not name:
                return None
            magnet_tag = page.select_one('a[href^="magnet:"]')
            if not magnet_tag:
                return None
            magnet = magnet_tag['href']
            size_text = ''
            seeders_text = '0'
            for li in page.select('ul.list li'):
                label = li.select_one('strong')
                value = li.select_one('span')
                if not label or not value:
                    continue
                label_text = label.get_text().strip().lower()
                if 'size' in label_text:
                    size_text = value.get_text().strip()
                    # Remove nested span (e.g. "(2,621,440,000 Bytes)")
                    size_text = re.sub(r'\(.*?\)', '', size_text).strip()
                elif 'seeders' in label_text:
                    seeders_text = value.get_text().strip()
            return {
                'name': name,
                'size': size_text or 'N/A',
                'size_gb': utils.parse_size(size_text),
                'seeders': utils.parse_seeders(seeders_text),
                'magnet': magnet,
                'provider': self.name,
            }

        with ThreadPoolExecutor(max_workers=8) as pool:
            futures = {pool.submit(fetch_detail, url): url for url in links[:20]}
            for future in as_completed(futures):
                try:
                    result = future.result()
                    if result:
                        results.append(result)
                except Exception:
                    pass
        return results
